### ISSUE TITLE
LinearRegression does not work with column-rank-deficient data

### ISSUE THREAD

**Issue by lukehutch** [haifengl/smile#88]
(Separating this discussion from #86.)

LinearRegression does not work with column-rank-deficient data, including datasets where all values in a column are zero, and presumably also for datasets where one column is linearly dependent on another. It throws an exception, saying that the data is rank-deficient (meaning column-rank-deficient, not row-rank-deficient.)

This is a bad problem for datasets containing rare features: in a given feature vector, for a specific k-fold cross validation division of the data, all the non-zero entries in the rare feature dimension can end up in the test set, and the training set will end up with all zeroes for that column, which will mean that LinearRegression will die for that fold. This could cause SMILE to crash only sometimes, on some random cross-validation divisions, and not on others. This is a problem because rare feature dimensions are quite commonly present in a model.

Both R and SciKit-Learn handle this situation fine, indicating that either they are using a different numerical method, or that they have special-case handling for this.

LASSO also doesn't work for column-rank-deficient data for a possibly different reason, giving NaN values as discussed in #86, but may be fixed after the changes discussed there. I have only tried LinearRegression, LASSO and GradientTreeBoost so far, so it's possible other regression and/or classification methods are also affected by this issue. (GradientTreeBoost works fine.) It would be nice if all the methods were verified to work in a sane way even given all-zero or linearly-dependent columns as input, since this situation is likely to come up with some frequency, particularly in datasets with rare features.

**Comment by alexeygrigorev** [comment]
+1 run onto this issue today as well

**Comment by haifengl** [comment]
By switching to SVD, we can solve this problem. QR requires full rank.

BTW, the rank of a matrix <= min(rows, cols). In least squares, we always have more rows than columns. When we say rand deficient in LS, we always means column rank deficient. see details at http://www.netlib.org/lapack/lug/node27.html

**Comment by haifengl** [comment]
I improve the code to handle constant columns in LASSo and ridge regression, which is in the master branch now. Before we release a new version, please git clone and build by yourself. Thanks!

**Comment by haifengl** [comment]
For linear regression, we definitely should use SVD instead of QR to handle rank deficient. It seems a lot of work to improve smile.matrix.

Originally I didn't expect people will pay attention to linear regression in smile. I put a lot of efforts to implement advanced algorithms for other learning methods (the motivation of creating smile). I was wrong :)

**Comment by lukehutch** [comment]
Personally, I always use linear regression as a baseline measurement. Any regressor should do at least as well as linear regression. It's easy then to see them how mutch lift you get from a given regression and set of parameters.

**Comment by haifengl** [comment]
@lukehutch it is a good practice. How's lasso performance comparison between smile and sklearn? Thanks!

**Comment by lukehutch** [comment]
(added comment to the other bug about speed of SMILE vs. sklearn with Lasso)

**Comment by haifengl** [comment]
I add another safe guard for qr method. When qr fails in OLS, it automatically falls back to SVD now.
